lib/Glimmer.pm

```diff
@@ -0,0 +1,494 @@
+		print STDERR $longorfs_err if $verbose;
+	    }
+	    
+	    if ($rc) {
+		if ($longorfs_err) {
+		    if ($longorfs_err =~ m/WHAT: ERROR:  No valid orfs found below entropy cutoff/so) {
+			#...Error is harmless --- ignore it
+		    }
+		    else {
+			die (qq(Could not extract training ORFs from contig $contig_id --- return-code $rc:\n), $longorfs_err);
+		    }
+		}
+		else {
+		    die qq(Could not extract training ORFs from contig $contig_id --- return-code $rc);
+		}
+	    }
+	    
+	    system("extract -t $tmp_contig $tmp_coords >> $tmp_train") 
+		&& die "Could not extract training sequences from contig $contig_id";
+	    
+	    #... Translate GLIMMER  $tmp_coords into SEED $training_tbl
+	    open(TMP_COORDS, "<$tmp_coords") || die "Could not read-open $tmp_coords";
+	    print TRAINING map { ++$orf_num;
+				 chomp $_;
+				 my (undef, $beg, $end) = split /\s+/o, $_;
+				 die "Bad coords in entry: $_" unless ($beg && $end);
+				 my $fid = qq(orf) . (qq(0)x(5-length($orf_num))) . $orf_num;
+				 my $loc = join(qq(_), ($contig_id, $beg, $end));
+				 $_ = qq($fid\t$loc\n)
+				 } <TMP_COORDS>;
+	}
+	close(TRAINING) || die "Could not close $training_tbl";
+	close(CONTIGS)  || die "Could not close $contigs_file";
+    }
+
+
+#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    print STDERR ("\n",
+		  "Training using:\n",
+		  "   contigs --- $training_contigs\n",
+		  "   ORFs    --- $training_tbl\n"
+		 )
+	if $verbose;
+    #-----------------------------------------------------------------------
+    open(CONTIGS, "<$training_contigs") 
+	|| die "Could not read-open $training_contigs";
+    
+    my ($len_of, $seq_of);
+    while (($contig_id, $seqP) = &read_fasta_record(\*CONTIGS))  {
+	$len_of->{$contig_id} = length($$seqP);
+	$seq_of->{$contig_id} = $$seqP;
+    }
+    close(CONTIGS) || die "Could not close $contigs_file";
+    
+    my $entry;
+    my $orf_num;
+    my $max_orf_num = 0;
+    my %training_tbl;
+    open(TBL,   "<$training_tbl") || die "Could not read-open $training_tbl";
+    open(TRAIN, ">$tmp_train")    || die "Could not write-open $tmp_train";
+    while (defined($entry = <TBL>)) {
+	chomp $entry;
+	my ($fid, $locus) = split /\t/, $entry;
+	
+	if ($fid =~ m/(\d+)$/) {
+	    $orf_num     = $1;
+	    $max_orf_num = ($orf_num > $max_orf_num) ? $orf_num : $max_orf_num;
+	}
+	else {
+	    die "Could not parse FID $fid for training entry $.: $entry";
+	}
+	
+	my $training_seq = qq();
+	my @exons        = split /,/, $locus;
+	foreach my $exon (@exons) {
+	    if ($exon =~ m/^(\S+)_(\d+)_(\d+)/) {
+		my ($contig, $beg, $end) = ($1, $2, $3);
+		
+		my $contig_seq = $seq_of->{$contig};
+		
+		my $dna = &get_dna_seq($contig, $beg, $end, $len_of, $seq_of);
+		
+		$training_seq .= lc($dna);
+	    }
+	    else {
+		die "Could not parse exon $exon for training entry $.: $entry";
+	    }
+	    
+	    my $training_ID = qq(orf) . (qq(0) x (5-length($orf_num))) . $orf_num;
```

Comment:
This is a super obscure way to write out a string -- if the training ID needs to be in a certain form, put a comment in your code about it, and then choose the simplest way to create the string. Even if this is an efficient way to write it, anyone coming to this code to fix bugs or add features is going to have to spend time working out what this is supposed to do.